### PR TITLE
feat(storage): add support for conditional headers to copy, and validate serialization

### PR DESCRIPTION
**Pull request by eppjame**
#### Description of changes

- Adds support for `x-amz-copy-source-if-match` and `x-amz-copy-source-if-unmodified-since` conditional headers
- Introduce header validation for all used headers
- Use existing URL validation to check endpoint construction
- Add basic integrity helpers `isNil` and `bothNilOrEqual` for convenience

#### Description of how you validated changes
- New unit tests covering all validations
- Manual testing in a sample application

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Unit Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [ ] Relevant documentation is changed or added (and PR referenced)

#### Checklist for repo maintainers

- [ ] Verify E2E tests for existing workflows are working as expected or add E2E tests for newly added workflows
- [ ] New source file paths included in this PR have been added to CODEOWNERS, if appropriate

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/storage/src/providers/s3/utils/client/utils/integrityHelpers.ts [hunk 1]
```diff
@@ -0,0 +1,10 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+export const isNil = <T>(value?: T) => {
```
Comment (scyrizales-amz) [inline packages/storage/src/providers/s3/utils/client/utils/integrityHelpers.ts:4]: I have found usages of lodash in amplify. Should we use lodash isNil?
`import isNil from 'lodash/isNil';`

Comment (eppjame) [reply packages/storage/src/providers/s3/utils/client/utils/integrityHelpers.ts:4]: I could see the argument if it was used in the storage module, but I think pulling in dependencies for each module would be decided individually.

Comment (scyrizales-amz) [reply packages/storage/src/providers/s3/utils/client/utils/integrityHelpers.ts:4]: got it

**Review by jimblanc — DISMISSED** [review]

File: packages/storage/__tests__/providers/s3/utils/client/S3/utils/integrityHelpers.test.ts [hunk 2]
```diff
@@ -0,0 +1,34 @@
+import {
+	bothNilOrEqual,
+	isNil,
+} from '../../../../../../../src/providers/s3/utils/client/utils/integrityHelpers';
+
+describe('isNil', () => {
```
Comment (ashika112) [inline packages/storage/__tests__/providers/s3/utils/client/S3/utils/integrityHelpers.test.ts:6]: nit: could use inline doc for explaining the util

File: packages/storage/src/providers/s3/types/index.ts [hunk 3]
```diff
@@ -17,8 +17,8 @@ export {
 	DownloadDataOptionsWithKey,
 	CopyDestinationOptionsWithKey,
 	CopySourceOptionsWithKey,
-	CopyWithPathSourceOptions,
```
Comment (ashika112) [inline packages/storage/src/providers/s3/types/index.ts:20]: this was intentional. we have a draft PR out to change this format. I would revert this naming change.

Comment (eppjame) [reply packages/storage/src/providers/s3/types/index.ts]: Done.

**Review by ashika112 — APPROVED** [review]

**Review by jimblanc — APPROVED** [review]

**Pull request merged**